python/ray/util/client/dataclient.py

```diff
@@ -262,35 +292,26 @@ def close(self) -> None:
     def _blocking_send(
         self, req: ray_client_pb2.DataRequest
     ) -> ray_client_pb2.DataResponse:
+        self.request_queue.put((req, None))
         with self.lock:
             self._check_shutdown()
-            req_id = self._next_id()
-            req.req_id = req_id
-            self.request_queue.put(req)
-            self.outstanding_requests[req_id] = req
 
-            self.cv.wait_for(lambda: req_id in self.ready_data or self._in_shutdown)
+            self.cv.wait_for(lambda: id(req) in self.ready_data or self._in_shutdown)
             self._check_shutdown()
 
-            data = self.ready_data[req_id]
-            del self.ready_data[req_id]
-            del self.outstanding_requests[req_id]
-            self._acknowledge(req_id)
+            resp = self.ready_data[id(req)]
+            del self.ready_data[id(req)]
+            del self.outstanding_requests[resp.req_id]
+            self._acknowledge(resp.req_id)
 
-        return data
+        return resp
 
     def _async_send(
         self,
         req: ray_client_pb2.DataRequest,
         callback: Optional[ResponseCallable] = None,
     ) -> None:
-        with self.lock:
-            self._check_shutdown()
-            req_id = self._next_id()
-            req.req_id = req_id
-            self.asyncio_waiting_data[req_id] = callback
-            self.outstanding_requests[req_id] = req
-            self.request_queue.put(req)
+        self.request_queue.put((req, callback))
```

Comment:
I think there is one thing different from before when the cluster is down.
Previously it'll throw an exception, but now, it'll be pushed to the queue. Is there anything we need to take care here?

Reply:
Which exception?

Reply:
let's add some comments here. so that future maintainers won't make the same mistake.

Reply:
Added a comment above `_check_shutdown()`. Let me know if anything else can be documented better.

Reply:
I mean comment like:
```
This function will be called from python GC and we should avoid using the lock to prevent a deadlock.
```

Reply:
Added. Btw I find it easier to raise an exception from _async_send() when the client has shut down.